server/service/src/service_provider.rs

```diff
@@ -80,9 +82,15 @@ impl ServiceProvider {
     }
 
     /// Creates a new service context with a new DB connection
-    pub fn context(&self) -> Result<ServiceContext, RepositoryError> {
+    pub fn context(
+        &self,
+        store_id: &str,
+        user_id: &str,
+    ) -> Result<ServiceContext, RepositoryError> {
```

Comment:
Probably a micro optimization but most of the time we already have a String for store_id and user_id and we could use this one instead of effectively cloning these strings...

A `base_context()`  without params could then do the initialization of the empty strings...

Reply:
We already have a `user_id: String` from the user and graphql could give us a `store_id: String` as well. So at least we have to clone the user_id into the context when using a `&str` argument.

However, I am fine leaving as it is don't think it will makes a big impact.

Reply:
Hmm, I think we would need to change quite a bit to use `String` from graphql input (as it's passed by reference/slice/`&sr` to graphql handlers).

I think the general rule for owned/referenced things (although this is somewhat personal stylistic preference):

* If only one method uses the input, use owned (1), otherwise (2) use reference and clone()/to_string() when it needs to be owned (3) further up the track. 

```rust
// 1
fn something(text: String) {
  something_else(text)
}
// 2
fn something_else(text: String) {
  another_one(&text)
  and_another(&text)
  // Could use just text here, but I think it's easier to extend if reference is used, i.e. what if another method is added that needs to use text, then it's easier to just another &text, vs cloning now to and_more
  and_more(&text)
}
// 3
fn and_another(text: &str) {
   // ...
   DataRow {
     id: uuid(),
     value: text.to_string()
   }
}
```

so ideally for our example (two clones are required, one in generate and one in log, could be reduced to 1 though)

```rust
fn graphql(store_id: String) {
  graphql_handler(store_id)
}

fn graphql_handler(store_id: String) {
   validate(/* .. */, &store_id);
   let ctx = service_provider.get_context(&store_id) // store as &str in context (although it does mean a bit more lifetime annotations)
}

fn service(ctx: &ServiceContext) {
   validate(ctx)
   generate(ctx)
   log(ctx)
}

fn generate(ctx: &ServiceContext) {
  DataRow {
     id: uuid(),
     store_id: ctx.store_id.to_string()
  }
}

fn log(ctx: &ServiceContext) {
   LogRow {
     id: uuid(),
     store_id: ctx.store_id.to_string()
  }
}
```

Anyways, conclusion 🤷, maybe ok to keep things as is, just because of the sheer amount of changes.

Reply:
Weep... [47ae359](https://github.com/openmsupply/open-msupply/pull/465/commits/47ae359f67d27733a7dc9aed0f23bd0227868212)